federation/pkg/federation-controller/namespace/namespace_controller.go

```diff
@@ -0,0 +1,353 @@
+	// add/update/delete operation on a namespace in ehtier federated API server or
+	// in some member of the federation.
+	namespaceDeliverer *util.DelayingDeliverer
+
+	// For triggering all namespaces reconcilation. This is used when
+	// a new cluster becomes available.
+	clusterDeliverer *util.DelayingDeliverer
+
+	// Contains namespaces present in members of federation.
+	namespaceFederatedInformer util.FederatedInformer
+	// For updating members of federation.
+	federatedUpdater util.FederatedUpdater
+	// Definitions of namespaces that should be federated.
+	namespaceInformerStore cache.Store
+	// Informer controller for namespaces that should be federated.
+	namespaceInformerController framework.ControllerInterface
+
+	// Stop chanels to stop the controler.
+	namespaceStopChan          chan struct{}
+	clusterReconcileStopChan   chan struct{}
+	namespaceReconcileStopChan chan struct{}
+}
+
+// A structure passed by delying deliver. It contains a namespace that should be reconciled and
+// the number of trials that were made previously and ended up in some kind of namespace-related
+// error (like failure to create).
+type namespaceItem struct {
+	namespace string
+	trial     int64
+}
+
+// NewNamespaceController returns a new namespace controller
+func NewNamespaceController(client federation_release_1_4.Interface) *NamespaceController {
+
+	// Builds stop chanels.
+	nc := &NamespaceController{
+		namespaceStopChan:          make(chan struct{}),
+		clusterReconcileStopChan:   make(chan struct{}),
+		namespaceReconcileStopChan: make(chan struct{}),
+	}
+
+	// Build delivereres for triggering reconcilations.
+	nc.namespaceDeliverer = util.NewDelayingDeliverer(make(chan *util.DelayingDelivererItem, namespaceChanSize))
+	nc.clusterDeliverer = util.NewDelayingDeliverer(make(chan *util.DelayingDelivererItem, clusterChanSize))
+
+	// Start informer in federated API servers on namespaces that should be federated.
+	nc.namespaceInformerStore, nc.namespaceInformerController = framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func(options api.ListOptions) (pkg_runtime.Object, error) {
+				return client.Core().Namespaces().List(options)
+			},
+			WatchFunc: func(options api.ListOptions) (watch.Interface, error) {
+				return client.Core().Namespaces().Watch(options)
+			},
+		},
+		&api.Namespace{},
+		controller.NoResyncPeriodFunc(),
+		framework.ResourceEventHandlerFuncs{
+			// Trigger reconcilation whenever some defiition is changed.
+			DeleteFunc: func(old interface{}) { nc.deliverNamespaceObj(old, 0, 0) },
+			AddFunc:    func(cur interface{}) { nc.deliverNamespaceObj(cur, 0, 0) },
+			UpdateFunc: func(old, cur interface{}) {
+				nold, _ := old.(*api_v1.Namespace)
+				ncur, _ := old.(*api_v1.Namespace)
+				// Object meta or spec updated. Status changes are ignored.
+				if !reflect.DeepEqual(nold.Spec, ncur.Spec) ||
+					!reflect.DeepEqual(nold.ObjectMeta, ncur.ObjectMeta) {
```

Comment:
I'm not convinced that this is what we would like to do. If e.g. namespace changes state to "Terminating" (which is status field change), it should trigger reconciliation. But it won't.

Also, on what ObjectMeta you would like to reconciile?